desc/compute/utils.py

```diff
@@ -11,45 +12,308 @@
+        otherwise, returns a dict of ndarray with keys for R_lmn, Z_lmn, etc.
+    """
+    deps = list(keys) + get_data_deps(*keys)
+    params = []
+    for key in deps:
+        params += data_index[key]["dependencies"]["params"]
+    params = sorted(list(set(params)))
+    if eq is None:
+        return params
+    params = {name: np.atleast_1d(getattr(eq, name)).copy() for name in params}
+    return params
+
+
+def get_transforms(*keys, eq, grid, **kwargs):
+    """Get transforms needed to compute a given quantity on a given grid.
+
+    Parameters
+    ----------
+    keys : str
+        Name of the desired quantity from the data index
+    eq : Equilibrium
+        Equilibrium to compute quantity for.
+    grid : Grid
+        Grid to compute quantity on
+
+    Returns
+    -------
+    transforms : dict of Transform
+        Transforms needed to compute key.
+        Keys for R, Z, L, etc
+
+    """
+    from desc.basis import DoubleFourierSeries
+    from desc.transform import Transform
+
+    derivs = get_derivs(*keys)
+    transforms = {"grid": grid}
+    for c in ["R", "L", "Z"]:
+        if c in derivs:
+            transforms[c] = Transform(
+                grid, getattr(eq, c + "_basis"), derivs=derivs[c], build=True
+            )
+    if "B" in derivs:
+        transforms["B"] = Transform(
+            grid,
+            DoubleFourierSeries(
+                M=kwargs.get("M_booz", 2 * eq.M),
+                N=kwargs.get("N_booz", 2 * eq.N),
+                NFP=eq.NFP,
+                sym=eq.R_basis.sym,
+            ),
+            derivs=derivs["B"],
+            build=True,
+            build_pinv=True,
+        )
+    if "w" in derivs:
+        transforms["w"] = Transform(
+            grid,
+            DoubleFourierSeries(
+                M=kwargs.get("M_booz", 2 * eq.M),
+                N=kwargs.get("N_booz", 2 * eq.N),
+                NFP=eq.NFP,
+                sym=eq.Z_basis.sym,
+            ),
+            derivs=derivs["w"],
+            build=True,
+            build_pinv=True,
+        )
+    return transforms
+
+
+def has_dependencies(qty, params, transforms, profiles, data):
+    """Determine if we have the ingredients needed to compute qty.
+
+    Parameters
+    ----------
+    qty : str
+        Name of something from the data index.
+    params : dict of ndarray
+        Dictionary of parameters we have.
+    transforms : dict of Transform
+        Dictionary of transforms we have.
+    profiles : dict of Profile
+        Dictionary of profiles we have.
+    data : dict of ndarray
+        Dictionary of what we've computed so far.
+
+    Returns
+    -------
+    has_dependencies : bool
+        Whether we have what we need.
+    """
+    return (
+        _has_data(qty, data)
+        and _has_params(qty, params)
+        and _has_profiles(qty, profiles)
+        and _has_transforms(qty, transforms)
+    )
+
+
+def _has_data(qty, data):
```

Comment:
slightly confusing to have this named "has_data" when it really is "_does_not_have_data" since it will return False if the data is present in the data dict already

Reply:
its really "has data dependencies" (ie all the things needed to compute x, and if we already have x there's no reason to compute it again) but that was too long.